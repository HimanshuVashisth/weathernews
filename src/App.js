import './App.css';
import React, { Component } from 'react';
import Button from './buttons';
import { DEFAULT_QUERY, PATH_BASE, PATH_SEARCH, PARAM_SEARCH, PARAM_PAGE } from './constants';

const largeColumn = { width: '40%' };
const midColumn = { width: '30%' };
const smallColumn = { width: '10%' };

class App extends Component {

    constructor(props) {
        super(props);

        this.state = {
            results: null,
            searchKey: '',
            searchValue: DEFAULT_QUERY,
            error: null,
        };

        this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
        this.setSearchTopStories = this.setSearchTopStories.bind(this);
        this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);
        this.onSearchSubmit = this.onSearchSubmit.bind(this);
        this.onDismiss = this.onDismiss.bind(this);

    }

    componentDidMount() {
        const { searchValue } = this.state;
        this.setState({ searchKey: searchValue });
        this.fetchSearchTopStories(searchValue);
    }

    fetchSearchTopStories(searchValue, page = 0) {
        fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchValue}&${PARAM_PAGE}${page}`)
            .then(response => response.json())
            .then(result => this.setSearchTopStories(result))
            .catch(e => this.setState({ error: e }));
    }

    setSearchTopStories(result) {
        const { hits, page } = result;
        const { searchKey, results } = this.state;

        const oldHits = results && results[searchKey] ? results[searchKey].hits : [];

        const updatedHits = [...oldHits, ...hits];

        this.setState({
            results: {
                ...results,
                [searchKey]: { hits: updatedHits, page }
            }
        })
    }

    onDismiss(id) {
        console.log('Inside onDismiss: ', JSON.stringify(id));
        const { searchKey, results } = this.state;
        const { hits, page } = results[searchKey];

        const isNotId = item => item.objectID !== id;
        const updatedHits = hits.filter(isNotId);
        this.setState({
            results: { ...results, [searchKey]: { hits: updatedHits, page } }
        });
    }

    onSearchChange(event) {
        this.setState({ searchValue: event.target.value });
    }

    onSearchSubmit(event) {
        const { searchValue } = this.state;
        this.setState({ searchKey: searchValue });

        if (this.needsToSearchTopStories(searchValue)) {
            this.fetchSearchTopStories(searchValue);
        }

        event.preventDefault();
    }

    needsToSearchTopStories(searchValue) {
        return !this.state.results[searchValue];
    }

    render() {

        const { searchValue, results, searchKey, error } = this.state;
        const page = (results && results[searchKey] && results[searchKey].page) || 0;
        const list = (results && results[searchKey] && results[searchKey].hits) || [];

        if (!results) { return null; }

        if (error) { return <p> Something went wrong. </p> }

        return (
            <div className="page" >
                <div className="interactions">
                    <Search value={searchValue} onChange={this.onSearchChange} onSubmit={this.onSearchSubmit}> Search </Search> <br />
                </div>

                { error ?
                    <div className="interactions">
                        <p>Something went wrong. </p>
                    </div>
                    : <Table list={list} onDismiss={this.onDismiss} />
                }

                { results &&
                    <Table list={list} onDismiss={this.onDismiss} />
                }

                <div className="interactions">
                    <Button onClick={() => this.fetchSearchTopStories(searchValue, page + 1)}> More </Button> <br />
                </div>
            </div>
        );
    }

}

export default App;

const Search = ({ value, onChange, onSubmit, children }) =>
    <form onSubmit={onSubmit}>
        <input type="text" value={value} onChange={onChange} />
        <button type="submit">
            {children}
        </button>
    </form>

const Table = ({ list, onDismiss }) =>
    <div className="table">
        {list.map(item =>
            <li>
                <div key={item.objectID} className="table-row">
                    <span style={{ largeColumn }}>
                        <a href={item.title}> {item.title}</a>
                    </span>
                    <span style={{ midColumn }}>{item.author}</span>
                    <span style={{ smallColumn }}>{item.num_comments}</span>
                    <span style={{ smallColumn }}>{item.points}</span>
                    <span style={{ smallColumn }}>
                        <Button onClick={() => onDismiss(item.objectID)} className="button-inline">
                            Dismiss
                        </Button>
                    </span>
                </div>
            </li>
        )}
    </div>


export {
    Button,
    Search,
    Table
}