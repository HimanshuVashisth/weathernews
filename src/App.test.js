import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import App, { Search, Button, Table } from './App';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {                         // Test Suite

  it('renders learn react link', () => {        // Test case [1 ... n] 
    const div = document.createElement('div');
    render(<App />, div);
  });

  test('has a valid snapshot', () => {          // Snapshot
    const component = renderer.create(<App />); // creates ${component.test.js.snap} file under __snapshots__ folder in project structure

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Search', () => {

  it('Search', () => {
    const div = document.createElement('div');
    render(<Search>Search</Search>, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Search>Search</Search>);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});


describe('Button', () => {

  it('Button', () => {
    const div = document.createElement('div');
    render(<Button>More Button</Button>, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Button>Button</Button>);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

describe('Table', () => {

  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 10001 },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 10002 },
    ],
  };

  it('Table', () => {
    const div = document.createElement('div');
    render(<Table {...props} />, div);
  });

  it('shows two items in the list', () => {
    const element = shallow(<Table {...props} />)
    expect(element.find('.table-row').length).toBe(2);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(<Table {...props} />);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
