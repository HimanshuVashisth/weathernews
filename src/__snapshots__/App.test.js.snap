// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `null`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  type="button"
>
  Button
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form>
  <input
    type="text"
  />
  <button
    type="submit"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <li>
    <div
      className="table-row"
    >
      <span
        style={
          Object {
            "largeColumn": Object {
              "width": "40%",
            },
          }
        }
      >
        <a
          href="1"
        >
           
          1
        </a>
      </span>
      <span
        style={
          Object {
            "midColumn": Object {
              "width": "30%",
            },
          }
        }
      >
        1
      </span>
      <span
        style={
          Object {
            "smallColumn": Object {
              "width": "10%",
            },
          }
        }
      >
        1
      </span>
      <span
        style={
          Object {
            "smallColumn": Object {
              "width": "10%",
            },
          }
        }
      >
        2
      </span>
      <span
        style={
          Object {
            "smallColumn": Object {
              "width": "10%",
            },
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Dismiss
        </button>
      </span>
    </div>
  </li>
  <li>
    <div
      className="table-row"
    >
      <span
        style={
          Object {
            "largeColumn": Object {
              "width": "40%",
            },
          }
        }
      >
        <a
          href="2"
        >
           
          2
        </a>
      </span>
      <span
        style={
          Object {
            "midColumn": Object {
              "width": "30%",
            },
          }
        }
      >
        2
      </span>
      <span
        style={
          Object {
            "smallColumn": Object {
              "width": "10%",
            },
          }
        }
      >
        1
      </span>
      <span
        style={
          Object {
            "smallColumn": Object {
              "width": "10%",
            },
          }
        }
      >
        2
      </span>
      <span
        style={
          Object {
            "smallColumn": Object {
              "width": "10%",
            },
          }
        }
      >
        <button
          className="button-inline"
          onClick={[Function]}
          type="button"
        >
          Dismiss
        </button>
      </span>
    </div>
  </li>
</div>
`;
